Linux porting work

dougvj opened the pull request:
Here is a first stab of implementing several backends for the linux port, most notably porting UI code to GTK+3, and abstracting the GL context creation. 

Currently the code compiles on Linux, but there are several linker errors preventing full compilation and preventing testing of the code. There is a lot of work left to be done, but it's progress.

I don't have access to a windows build environment so I need these changes tested for that platform as well.

janisozaur commented on src/xenia/base/filesystem_posix.cc:
```diff
@@ -0,0 +1,193 @@
+}
+
+static uint64_t convertUnixtimeToWinFiletime(uint64_t unixtime) {
+    uint64_t filetime = filetime = (unixtime * 10000000) + 116444736000000000;
```
some magic number, perhaps add a comment what it is?

dougvj replied:
done

janisozaur commented:
https://ci.appveyor.com/project/benvanik/xenia/build/1.0.337#L432:
```
..\src\xenia\ui\gl\gl_context_win.cc(10): fatal error C1083: Cannot open include file: 'xenia/ui/gl/glx_context.h': No such file or directory [C:\dev\xenia\build\xenia-ui-gl.vcxproj]
```

Razzile commented:
![image](https://cloud.githubusercontent.com/assets/5256381/15212443/89971dca-1838-11e6-8008-8bf78d54624a.png)


jokes aside I'm glad to see some linux work 👍

darkf commented:
IMHO this should be split into two parts if possible (one for the GTK+ stuff, one for the compiling on Linux stuff), if that makes sense to do.

Secondly, your license headers have the wrong names and dates.

darkf commented:
If you're going to leave some of that stuff unimplemented (like the socket stuff) you should at least make it emit an unimplemented warning.

dougvj commented:
I noticed there are some build errors on wgl_context_win.cc. I will tackle that after work. I will try to setup a build environment for windows so I don't have to play the commit and pray game.

sephiroth99 commented:
NIce! I had started a while back, but I stopped working on it seriously. Anyway I put my stuff here https://github.com/sephiroth99/xenia/tree/linux , it's mostly dummy-ing out the stuff for it to build on Linux and with some fixes.

The linking error is an issue with premake5. For the linking to work, premake needs to add a link group for all the libraries. It added support for gcc in this commit https://github.com/premake/premake-core/pull/423 but if you use clang you need to modify premake (easy to see in that commit how to change for clang).

Anyway can't wait to try it out!

DrChat commented:
For now, you can stub out most of xam_net.cc. I've got a lot of local changes there to revamp it, just never committed them yet.

DrChat commented on src/xenia/ui/gl/gl_context_win.cc:
```diff
@@ -0,0 +1,297 @@
+ ******************************************************************************
+ */
+
+#include "xenia/ui/gl/gl_context.h"
```
This should be gl_context_win.h

DrChat commented on src/xenia/ui/gl/gl_context_x11.cc:
```diff
@@ -0,0 +1,317 @@
+DEFINE_bool(disable_gl_context_reset, false,
+            "Do not aggressively reset the GL context (helps with capture "
+            "programs such as OBS or FRAPS).");
+DEFINE_bool(gl_debug, false, "Enable OpenGL debug validation layer.");
```
Move these to gl_context.cc and put DECLARE_bools in its header.

Margen67 commented:
bump

vgturtle127 commented:
Is GTK a good choice? I thought it was being mostly retired due to code complexity and other issues. Maybe I am wrong, but I thought that's what I heard. 😆

John-Gee commented:
GTK3 is still maintained and GTK4 is getting there so no GTK is not getting retired.

vgturtle127 commented:
@John-Ge Alright, good to know.

scribam commented:
Is anyone currently working on it ?

dougvj commented:
I am still around, I have been distracted with other projects. Hopefully I will get back around to this in the next month or two.

scribam commented:
Thank you for the update.
Would you mind adding the package dependencies in the .travis.yml file?

dougvj commented:
xenia-app compiles and links now. There are some runtime checks that fail when creating the GTK menu bar, I will debug that sometime this week. 

I merged and rebased my changes from a year ago with the master repo, cleaning up the commit history in the process. I added some commits from @sephiroth99 who has also been working on this. I then implemented several changes including vulkan instantiation and the posix filesystem stuff. 

Right now the biggest thing that probably needs to be implemented from this point, aside from general testing and debugging, are the higher level synchronization constructs (Event, Semaphore, Timer, etc). I only stubbed those out. 

@scribam I will also update the .travis.yml stuff soon. My current build environment is not Ubuntu, but I have a VM I can use to make sure things are working properly

tmccaff commented:
OpenGL is no longer supported no reason to add that.

tmccaff commented:
Fact most likely OpenGL will be completely deleted soon once Vulkan issues fixed
